java/ql/src/Security/CWE/CWE-378/TempDirHijackingVulnerability.ql

```diff
@@ -0,0 +1,427 @@
+  )
+}
+
+/**
+ * Holds for any guard `g` that is itself guarded by a boolean variable access guard.
+ *
+ * For example, the following code: `if (isDirectory && !file.mkdir()) { ... }`.
+ */
+private predicate booleanVarAccessGuardsGuard(Guard g) {
+  exists(Guard g2 | g2 = any(VarAccess va | va.getType() instanceof BooleanType) |
+    g2.controls(g.getBasicBlock(), true)
+  )
+}
+
+/**
+ * Gets any `MethodAccess` that access string returning methods on the expression `e` of type `File`.
+ */
+private MethodAccess getStringAccessOnFile(Expr e) {
+  result =
+    any(MethodAccess fileMethodAccess |
+      fileMethodAccess.getMethod().getDeclaringType() instanceof TypeFile and
+      fileMethodAccess.getQualifier() = e and
+      fileMethodAccess.getMethod().getReturnType() instanceof TypeString
+    )
+}
+
+private Argument getThrowableConstructorParam() {
+  result =
+    any(Argument a |
+      exists(ConstructorCall c |
+        c.getConstructor().getDeclaringType().getAnAncestor() instanceof TypeThrowable and
+        c.getAnArgument() = a
+      )
+    )
+}
+
+private predicate safeUse(Expr e) {
+  exists(AndLogicalExpr andExp |
+    andExp.getType() instanceof BooleanType and andExp.getAnOperand() = e
+  )
+  or
+  exists(AssignAndExpr assignAndExp |
+    assignAndExp.getType() instanceof BooleanType and assignAndExp.getSource() = e
+  )
+  or
+  // File is being concatenated to a string, probably for a log or exception message
+  exists(AddExpr addExp |
+    addExp.getType() instanceof TypeString and
+    (
+      addExp.getRightOperand() = e
+      or
+      // A method call, like `File.getAbsolutePath()` is being called and concatenated to the end of a a string
+      addExp.getRightOperand() = getStringAccessOnFile(e)
+      or
+      // A method call, like `File.getAbsolutePath()` is being called and prepended to another string with a leading space character.
+      addExp.getLeftOperand() = getStringAccessOnFile(e) and
+      addExp.getRightOperand().(CompileTimeConstantExpr).getStringValue().matches(" %")
+    )
+  )
+  or
+  // File is being used to construct an exception message
+  e = getThrowableConstructorParam()
+  or
+  // A call to `File::deleteOnExit`
+  exists(MethodAccess ma |
+    ma.getMethod().hasName("deleteOnExit") and
+    ma.getMethod().getDeclaringType() instanceof TypeFile and
+    ma.getQualifier() = e
+  )
+  or
+  // An assignment to a variable that is exclusively used when safe
+  e = any(VariableAssign assign | safeUse(assign.getDestVar().getAnAccess())).getSource() and
+  not e = any(VariableAssign assign | not safeUse(assign.getDestVar().getAnAccess())).getSource()
+  or
+  // Data flow exists exclusively to locations that are known to be safe
+  DataFlow::localExprFlow(getStringAccessOnFile(e), any(Expr sink | safeUse(sink))) and
+  not DataFlow::localExprFlow(getStringAccessOnFile(e), any(Expr sink | not safeUse(sink)))
```

Comment:
Let's start over with this: could you set out what circumstance you want this `safeUse` / flow-to-unsafe-use stuff to recognise?

Reply:
Unsafeness is very broadly defined here -- since you're currently using `not safeUse` as a global dataflow sink, could you just remove the transitory "method return" and "argument call" cases and look explicitly for the final 4 bullet points that you regard as unsafe? Checking for the existence of an unsafe use will be much cheaper than trying to establish that all uses are safe.

The other use of `safeUse` is in trying to define a barrier guard, but would it be enough to say that checking `s.mkdir()`'s return value is a barrier-guard on all uses of `s` that it controls, regardless of whatever we think about `s` so far?